Getting started with the Content Delivery API (#5131)
umbraco/UmbracoDocs: 12/umbraco-cms/content-delivery-api.md

---
description: Get started with the Content Delivery API.
---

# Content Delivery API

## Overview

> With its several extension points, this new API delivers headless capabilities built directly into Umbraco. It allows you to retrieve your content items in a JSON format and lets you preset them in different channels, using your preferred technology stack. This feature preserves the friendly editing experience of Umbraco, while also ensuring a performant delivery of content, even in headless scenarios.

Reviewer: Maybe not start the entire article with the mention of extension points; we'd like people to understand that they can use the API out of the box.

```suggestion
The Content Delivery API delivers headless capabilities built directly into Umbraco. It allows you to retrieve your content items in a JSON format and lets you preset them in different channels, using your preferred technology stack. This feature preserves the friendly editing experience of Umbraco, while also ensuring a performant delivery of content in a headless fashion. And with its several extension points, you can tailor this API to fit a broad range of requirements.
```

### Register the Content Delivery API dependencies

In the `Startup.cs` file, register the API dependencies in the `ConfigureServices` by adding `.AddDeliveryApi()`:

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddUmbraco(_env, _config)
        .AddBackOffice()             
        .AddWebsite()
        // Register all Delivery API dependencies
        .AddDeliveryApi()
        .AddComposers()
        .Build();
}
```

### Enable the Content Delivery API

Add the `DeliveryApi` configuration section in `appsettings.json` and set the value of the `Enabled` key to `true`:

{% code title="appsettings.json" %}
```json
{
    "Umbraco": {
        "CMS": {
            "DeliveryApi": {
                "Enabled": true
            }
        }
    }
}
```
{% endcode %}

### Additional configuration

> Once the Delivery API has been configured on your project, you need to be aware that all your content will be made available to the public. However, there are a few additional configuration options that you can use to restrict access to the Delivery API endpoints and limit the content that is returned.

Reviewer: ```suggestion
Once the Delivery API has been configured on your project, all your published content will be made available to the public by default. However, there are a few additional configuration options that you can use to restrict access to the Delivery API endpoints and limit the content that is returned.
```

{% code title="appsettings.json" %}
```json
{
    "Umbraco": {
        "CMS": {

>             "DeliveryApi": {
>                 "Enabled": true,
>                 "PublicAccess": true,
>                 "ApiKey": "my-api-key",
>                 "DisallowedContentTypeAliases": ["alias1", "alias2", "alias3"]

Reviewer: See also comment about `"RichTextOutputAsJson"` below.

```suggestion
            "DeliveryApi": {
                "Enabled": true,
                "PublicAccess": true,
                "ApiKey": "my-api-key",
                "DisallowedContentTypeAliases": ["alias1", "alias2", "alias3"],
                "RichTextOutputAsJson": false
```

            }
        }
    }
}
```
{% endcode %}

> * `Umbraco.CMS.DeliveryApi.PublicAccess` determines whether the Delivery API (_if enabled_) should be publicly accessible or if access should require an API key.
> * `Umbraco.CMS.DeliveryApi.ApiKey` specifies what API key to use for authorizing access to the API when public access is disabled. This setting is also used for managing access to draft content.
> * `Umbraco.CMS.DeliveryApi.DisallowedContentTypeAliases` contains the aliases of the content types that would never be exposed through the Delivery API.

Reviewer: The appsettings notation uses colon to separate section paths. Alternatively one could leave out the `Umbraco:CMS:DeliveryApi` part entirely?

```suggestion
* `Umbraco:CMS:DeliveryApi:PublicAccess` determines whether the Delivery API (_if enabled_) should be publicly accessible or if access should require an API key.
* `Umbraco:CMS:DeliveryApi:ApiKey` specifies the API key to use for authorizing access to the API when public access is disabled. This setting is also used for accessing draft content for preview.
* `Umbraco:CMS:DeliveryApi:DisallowedContentTypeAliases` contains the aliases of the content types that should never be exposed through the Delivery API, regardless of all other configurations.
```

> 

Reviewer: We need to document the `"RichTextOutputAsJson"` setting too. At this point we probably don't need to dive too deep into the actual functionality, so maybe something like

```suggestion
* `Umbraco:CMS:DeliveryApi:RichTextOutputAsJson` enables outputting rich text content as JSON rather than the default HTML output. JSON can be a preferred format in many scenarios, not least because it supports routing of internal links better than HTML does.
```

Author: I decided to leave it out of this section because it is meant to cover "_additional configuration options that you can use to **restrict access** to the Delivery API endpoints and **limit the content** that is returned_" (_as per the introduction to the section_). We can either write a transition sentence after the three "original" settings and then mention that it also exists or leave it for another article where demonstrate how to interact with the endpoints 🤔 

What would you prefer?

Reviewer: Yeah I gathered as much. But the setting will be there by default (as false) for a newly created V12 project, so we should document it while we're documenting the others. A transition sentence would be great 👍

Author: Alright, I will add it then 👍

{% hint style="info" %}

> To test the functionality of the API, you would need to create some content items first.

Reviewer: ```suggestion
To test the functionality of the API, you need to create some content first.
```

{% endhint %}

## Endpoints

> The output produced by the Delivery API can either represent a specific content item or a paged list of multiple items. Before exploring the endpoints that API has to offer, there are a few concepts to keep in mind.

Reviewer: ```suggestion
The output produced by the Delivery API can either represent a specific content item or a paged list of multiple items. Before exploring the API endpoints, there are a few concepts to keep in mind.
```

### Concepts

<details>

> <summary>Content item JSON Schema</summary>

Reviewer: We don't have an actual JSON schema yet 😄 

```suggestion
<summary>Content item JSON structure</summary>
```

> The Delivery API uses the JSON schema below to represent the retrieved content items, which consist of a range of properties. Basic JSON properties that any item returned by the API will contain include `name`, `id`, and t_ype_ (`contentType` in this case). All properties added to the content item can be found under the `properties` field.  Depending on the used property editor, the value associated with it in the JSON response can be a _string_, _number_, _boolean expression_, _array_, _object_ or _`null`_ (when a picker property is empty). The `route` property provides information about the `path` of the content item, as well as details about the root node value that is represented by the `startItem` object. We will discuss the concept of a `startItem` in more detail in the next section. Finally, if the content item varies by culture, the `cultures` property will contain information about all configured cultures for the content node, including the culture variant `path` and `startItem` for each one.

Reviewer: ```suggestion
The Delivery API outputs the JSON structure outlined below to represent the retrieved content items, which consist of a range of properties:

* Basic properties for any content item include `name`, `id` and `contentType`.
* All editorial properties from the content type can be found in the `properties` collection.  Depending on the configured property editor, the property output value can be a _string_, _number_, _boolean expression_, _array_, _object_ or _`null`_. 
* The `route` property provides the `path` of the content item, as well as details about the root node value that is represented by the `startItem` object. We will discuss the concept of a `startItem` in more detail in the next section.
* If the content item varies by culture, the `cultures` property will contain information about all configured cultures for the content node, including the culture variant `path` and `startItem` for each one.
```

```json

> {
>   "name": "string",
>   "route": {
>     "path": "string",
>     "startItem": {
>       "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
>       "path": "string"
>     }
>   },
>   "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
>   "contentType": "string",
>   "properties": {
>     "property1Alias": "string",
>     "property2Alias": 0,
>     "property3Alias": true,
>     "property4Alias": [],
>     "property5Alias": {},
>     "property6Alias": null
>   },
>   "cultures": {
>     "cultureIsoCode1": {
>       "path": "string",
>       "startItem": {
>         "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
>         "path": "string"
>       }
>     },
>     "cultureIsoCode2": {
>       "path": "string",
>       "startItem": {
>         "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
>         "path": "string"
>       }
>     }
>   }
> }

Reviewer: To make the point clearer, don't reuse the GUID between the start item and the actual item. They can of course be equal if one is requesting the root item, but... just to make a point 😄 

```suggestion
{
  "name": "string",
  "route": {
    "path": "string",
    "startItem": {
      "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
      "path": "string"
    }
  },
  "id": "11fb598b-5c51-4d1a-8f2e-0c7594361d15",
  "contentType": "string",
  "properties": {
    "property1Alias": "string",
    "property2Alias": 0,
    "property3Alias": true,
    "property4Alias": [],
    "property5Alias": {},
    "property6Alias": null
  },
  "cultures": {
    "cultureIsoCode1": {
      "path": "string",
      "startItem": {
        "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
        "path": "string"
      }
    },
    "cultureIsoCode2": {
      "path": "string",
      "startItem": {
        "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
        "path": "string"
      }
    }
  }
}
```

In Umbraco, `Umbraco.CMS.Global.HideTopLevelNodeFromPath` setting can be used to enable or disable including the top-level node in the URL of your content items. If the setting is set to `false`, you can still use the root node URL segment in the `path` parameter of the "_Get by route_" endpoint. However, the preferred approach is to make use of the `Start-Item` header to specify the URL segment or `id` of the root content node. This way, you can omit the root node URL segment from the content item path when querying for a content item. The `HideTopLevelNodeFromPath` setting does not affect the output URLs when rendering the content output.

</details>

<details>

<summary>Output expansion</summary>

> **Output expansion** is a concept that refers to the ability to retrieve additional data about related content in the JSON output. By default, when a property that allows picking a different content item is used in a content node, for example, through a content picker property editor, the JSON representation of that content node will only contain a shallow representation of the referenced item. That means, only the basic information about the linked item, without the nested properties. However, with output expansion, it is possible to expand all or specific properties of the content, provided they are expandable and referenced at the first level.
> 
> This feature can be used when querying for a content item or items through the Delivery API, by adding an `expand` parameter to the query. The value of this parameter can be either `"all"` to expand all properties on the content item, or `"property:alias, alias, alias"` to expand specific ones.
> 
> 

Reviewer: ```suggestion
**Output expansion** allows you to retrieve additional data about related content in the API output for a given content item. 

By default, a content property that allows picking a different content item (for example a content picker property) outputs a shallow representation of the picked item. That means, only the basic information about the picked item, without the item properties. However, with output expansion, it is possible to include the properties of the picked item in the API output.

This feature can be used both when querying for single and multiple content items, by adding an `expand` parameter to the query. The value of this parameter can be either `"all"` to expand all properties on the requested content item, or `"property:alias, alias, alias"` to expand specific ones.

```

> The following snippet demonstrates the output of a "My post" content node without expanding the `linkedItem` content picker property.

Reviewer: ```suggestion
The following snippet demonstrates the default output of a content item (without expanding any properties).
```

> 

Reviewer: _The "suggest" feature does not like embedding code... but I'd suggest the same structure as below, without the expansion - just to show the request._

#### Request

```
GET
/umbraco/delivery/api/v1/content/item/9bdac0e9-66d8-4bfd-bba1-e954ed9c780d
```

#### Response

Author: So to have that structure as well before the default output?

Reviewer: Yep 😄

            "properties": {
                "title": "My demo blog",
                "description": "Nihil incididunt dolores adipisicing placeat quisque imperdiet interdum autem, dolorem fusce rhoncus sunt leo inventore dictumst quisque, voluptatem, magni justo nostrud deserunt! Natus ipsam commodi dignissimos, sodales ab.\n"
            }
        }
    },
    "cultures": {}
}
```
{% endcode %}

The built-in property editors in Umbraco that allow for output expansion are:  &#x20;

* `Umbraco.ContentPicker`
* `Umbraco.MultiNodeTreePicker`

</details>

<details>

<summary>Preview</summary>

Similar to the preview concept in Umbraco, the Delivery API allows for requesting unpublished content through its endpoints. This can be done by setting a `Preview` header to `true` in the API request. However, accessing draft versions of your content nodes requires authorization via an API key configured in `appsettings.json` file - `Umbraco:CMS:DeliveryApi:ApiKey` setting. To obtain preview data, you must add the `Api-Key` request header containing the configured API key to the appropriate endpoints. Draft content is not going to be included in the JSON response otherwise.

> 

Reviewer: Would be nice with a request sample here that includes both the `Preview` and the `Api-Key` headers. Something like the request sample in the start item section.

</details>

<details>

<summary>Localization</summary>

If your content is available in multiple languages, the Delivery API can resolve localized content. When querying content by `id`, the `Accept-Language` header can be used to request variant content. When querying content by path, the culture is already known and included in the path, making the `Accept-Language` header unnecessary. However, if the header is present in the request, its value will take precedence over any other configuration settings. Localization is also supported by the means of the CMS's culture and hostname configuration.

> 

Reviewer: Would be nice with a request sample here that includes the `Accept-Language` header. Something like the request sample in the start item section.

---
description: Get started with the Content Delivery API.
---

# Content Delivery API

> ## Overview

Reviewer 2: Do we need this heading? Seems a bit redundant, like, we do not really need to know that this is an overview 🤔 
What do you think?

Author: I agree 😊 I'll remove it

> The Content Delivery API delivers headless capabilities built directly into Umbraco. It allows you to retrieve your content items in a JSON format and lets you preset them in different channels, using your preferred technology stack. This feature preserves the friendly editing experience of Umbraco, while also ensuring a performant delivery of content in a headless fashion. And with its several extension points, you can tailor this API to fit a broad range of requirements.

Reviewer 2: ```suggestion
The Content Delivery API delivers headless capabilities built directly into Umbraco. It allows you to retrieve your content items in a JSON format and lets you preset them in different channels, using your preferred technology stack. This feature preserves the friendly editing experience of Umbraco, while also ensuring a performant delivery of content in a headless fashion. And with its different extension points, you can tailor this API to fit a broad range of requirements.
```

## Getting Started

> When upgrading an existing project to version 12, you would need to opt-in explicitly for using the Delivery API. Below you will find the steps you need to take in order to configure it for your Umbraco project. If you start with a fresh Umbraco 12 installation, the Delivery API will be enabled by default and therefore you can skip straight to the [Additional configuration](content-delivery-api.md#additional-configuration) section.

Reviewer 2: ```suggestion
When upgrading an existing project to Umbraco 12, you will need to opt-in explicitly for using the Delivery API. Below you will find the steps you need to take in order to configure it for your Umbraco project.

When you start with a fresh Umbraco 12 installation, the Delivery API will be enabled by default. This means that you can skip straight to the [Additional configuration](content-delivery-api.md#additional-configuration) section.
```

### Register the Content Delivery API dependencies

> In the `Startup.cs` file, register the API dependencies in the `ConfigureServices` by adding `.AddDeliveryApi()`:

Reviewer 2: Let's use steps here instead!

```suggestion
1. Open your projects `Startup.cs` file.'
2. Register the API dependencies in the `ConfigureServices` method by adding `.AddDeliveryApi()`:
```

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddUmbraco(_env, _config)
        .AddBackOffice()             
        .AddWebsite()
        // Register all Delivery API dependencies
        .AddDeliveryApi()
        .AddComposers()
        .Build();
}
```

### Enable the Content Delivery API

Add the `DeliveryApi` configuration section in `appsettings.json` and set the value of the `Enabled` key to `true`:

{% code title="appsettings.json" %}
```json
{
    "Umbraco": {
        "CMS": {
            "DeliveryApi": {
                "Enabled": true
            }
        }
    }
}
```
{% endcode %}

### Additional configuration

> Once the Delivery API has been configured on your project, all your published content will be made available to the public by default. However, there are a few additional configuration options that you can use to restrict access to the Delivery API endpoints and limit the content that is returned.

Reviewer 2: ```suggestion
Once the Delivery API has been configured on your project, all your published content will be made available to the public by default. A few additional configuration options allow you to restrict access to the Delivery API endpoints and limit the content that is returned.
```

    }
  },
  "id": "11fb598b-5c51-4d1a-8f2e-0c7594361d15",
  "contentType": "string",
  "properties": {
    "property1Alias": "string",
    "property2Alias": 0,
    "property3Alias": true,
    "property4Alias": [],
    "property5Alias": {},
    "property6Alias": null
  },
  "cultures": {
    "cultureIsoCode1": {
      "path": "string",
      "startItem": {
        "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
        "path": "string"
      }
    },
    "cultureIsoCode2": {
      "path": "string",
      "startItem": {
        "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
        "path": "string"
      }
    }
  }
}
```

</details>

<details>

<summary>Start item</summary>

When working headlessly in a multi-site setup, sometimes it can be difficult to determine the site context of a particular content item. This is where the _start item_ comes into play.

The start item represents the root of a content item and is commonly returned from the API in conjunction with a content path. For instance:

```json
{
  "route": {
    "path": "/articles/2023/getting-started",
    "startItem": {
      "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
      "path": "docs-portal"
    }
  }
}
```

> This means that the content item resides under the root node with "docs-portal" as an alias and can be retrieved using the path "/articles/2023/getting-started".

Reviewer 2: ```suggestion
This means that content items that reside under the root node with `docs-portal` as an alias can be retrieved using via `/articles/2023/getting-started`.
```

> The start item is also quite helpful when obtaining content from the API. By supplying either the start item `id` or `path` in the `Start-Item` header, you can instruct the Delivery API to use the corresponding root node as the starting point for the requested content operation:

Reviewer 2: I tried to shorten the sentence. Didn't succeed. Can you give it a try? Perhaps it could be split into two?

```suggestion
The start item is helpful when obtaining content from the API. Supplying the start item `id` or `path` in the `Start-Item` header, allows you to instruct the Delivery API to use the corresponding root node as the starting point:
```

Author: Maybe something like this:
> The start item can also be quite helpful through the `Start-Item` request header when obtaining content from the Delivery API. Supplying a root node `id` or `path` as the header value, allows you to specify which is the starting point for the requested content operation:

OR 

> The start item is also quite helpful when obtaining content from the Delivery API by applying the `Start-Item` request header. Supplying a root node `id` or `path` as the header value, allows you to specify which is the starting point for the requested content operation:

```http
GET /umbraco/delivery/api/v1/content/item/articles/2023/getting-started
Start-Item: docs-portal
```

</details>

<details>

<summary>Output expansion</summary>

**Output expansion** allows you to retrieve additional data about related content in the API output for a given content item.

> By default, a content property that allows picking a different content item (for example a content picker property) outputs a shallow representation of the picked item. That means, only the basic information about the picked item, without the item properties. However, with output expansion, it is possible to include the properties of the picked item in the API output.

Reviewer 2: ```suggestion
By default, a content property that allows picking a different content item (like a content picker property) outputs a shallow representation of the item. That means, only the basic information about the picked item, without the item properties. However, with output expansion, it is possible to include the properties of the picked item in the API output.
```

> This feature can be used both when querying for single and multiple content items, by adding an `expand` parameter to the query. The value of this parameter can be either `"all"` to expand all properties of the requested content item or `"property:alias, alias, alias"` to expand specific ones.

Reviewer 2: ```suggestion
This feature can be used when querying for both single and multiple content items, by adding an `expand` parameter to the query. The value of this parameter can be either `"all"` to expand all properties of the requested content item or `"property:alias, alias, alias"` to expand specific ones.
```

> The following JSON snippet demonstrates the default output of a content item (without expanding any properties):

Reviewer 2: There are two snippets, right? 🤔 

```suggestion
The following JSON snippets demonstrate the default output of a content item (without expanding any properties).
```

Author: Only the one right after shows "_default output of a content item (**without** expanding any properties)_". The second one will expand the properties. So "snippet" should be singular here 🙂

        "linkedItem": {
            "name": "Demo blog",
            "route": {
                "path": "/demo-blog/",
                "startItem": {
                    "id": "5d5ae914-9885-4ee0-a14b-0ab57f501a55",
                    "path": "demo-blog"
                }
            },
            "id": "5d5ae914-9885-4ee0-a14b-0ab57f501a55",
            "contentType": "blog",
            "properties": {}
        }
    },
    "cultures": {}
}
```
{% endcode %}

> And here is an example of how an expanded representation might look for the `linkedItem` property that references another content item with properties `title` and `description`:

Reviewer 2: ```suggestion
The following is an example of how an expanded representation might look for the `linkedItem` property that references another content item with properties `title` and `description`:
```

Requested culture
{% endswagger-parameter %}

{% swagger-parameter in="header" name="Api-Key" type="String" required="false" %}
Access token
{% endswagger-parameter %}

{% swagger-parameter in="header" name="Preview" type="Boolean" required="false" %}
Whether draft content is requested
{% endswagger-parameter %}

{% swagger-parameter in="header" name="Start-Item" type="String" required="false" %}
URL segment or GUID of the root content item
{% endswagger-parameter %}

{% swagger-parameter in="query" name="expand" type="String" required="false" %}
Which properties to expand and therefore include in the output if they refer to another piece of content
{% endswagger-parameter %}

{% swagger-response status="200: OK" description="Paginated list of content items" %}

{% endswagger-response %}

{% swagger-response status="400: Bad Request" description="Invalid request" %}

{% endswagger-response %}
{% endswagger %}

## Feature Capabilities

* Querying for content item(s)
* Localization support
* Output expansion
* Preview
* Multi-site support
* API authorization
* Denylist of document types
* Custom property editors

## Current Limitations

> While the Content Delivery API provides a powerful and flexible way to retrieve content from the Umbraco CMS, there are certain limitations to be aware of. In this section, we'll discuss some of the known limitations of the API, and how to work around them if necessary.

Reviewer 2: ```suggestion
The Content Delivery API provides a powerful and flexible way to retrieve content from the Umbraco CMS. There are, however, certain limitations to be aware of. 

In this section, we will discuss some of the known limitations of the API, and how to work around them if necessary.
```

This feature can be used when querying for both single and multiple content items, by adding an `expand` parameter to the query. The value of this parameter can be either `"all"` to expand all properties of the requested content item or `"property:alias, alias, alias"` to expand specific ones.

The following JSON snippet demonstrates the default output of a content item (without expanding any properties).

**Request**

```http
GET /umbraco/delivery/api/v1/content/item/9bdac0e9-66d8-4bfd-bba1-e954ed9c780d
```

**Response**

> {% code title="Shallow output for " %}

Author: The title seems incomplete that way

```
{% endcode %}

Below is an example of how an expanded representation might look for the `linkedItem` property that references another content item with properties `title` and `description`:

**Request**

```http
GET /umbraco/delivery/api/v1/content/item/9bdac0e9-66d8-4bfd-bba1-e954ed9c780d?expand=property:linkedItem
```

**Response**

> {% code title="Expanded output for " %}

Author: Same here

## Feature Capabilities

> * the Querying for the content item(s)

Author: Is this change intended?
